miner/src/block_assembler.rs

```diff
@@ -429,80 +433,41 @@ impl<CS: ChainStore + 'static> BlockAssembler<CS> {
 
     fn prepare_uncles(
         &self,
-        tip: &Header,
         current_epoch_ext: &EpochExt,
-    ) -> (Vec<UncleBlock>, Vec<H256>) {
-        let max_uncles_age = self.shared.consensus().max_uncles_age();
-        let mut excluded = FnvHashSet::default();
-
-        // cB
-        // tip      1 depth, valid uncle
-        // tip.p^0  ---/  2
-        // tip.p^1  -----/  3
-        // tip.p^2  -------/  4
-        // tip.p^3  ---------/  5
-        // tip.p^4  -----------/  6
-        // tip.p^5  -------------/
-        // tip.p^6
-        let mut block_hash = tip.hash().to_owned();
-        excluded.insert(block_hash.clone());
-        for _depth in 0..max_uncles_age {
-            if let Some(block) = self.shared.block(&block_hash) {
-                excluded.insert(block.header().parent_hash().to_owned());
-                for uncle in block.uncles() {
-                    excluded.insert(uncle.header.hash().to_owned());
-                }
-
-                block_hash = block.header().parent_hash().to_owned();
-            } else {
-                break;
-            }
-        }
-
-        let current_number = tip.number() + 1;
-
+        candidate_uncles: &mut LruCache<H256, Arc<Block>>,
+    ) -> FnvHashSet<UncleBlock> {
+        let store = self.shared.store();
+        let epoch_number = current_epoch_ext.number();
         let max_uncles_num = self.shared.consensus().max_uncles_num();
-        let mut included = FnvHashSet::default();
-        let mut uncles = Vec::with_capacity(max_uncles_num);
-        let mut bad_uncles = Vec::new();
+        let mut uncles = FnvHashSet::with_capacity_and_hasher(max_uncles_num, Default::default());
 
-        for (hash, block) in self.candidate_uncles.iter() {
-            if uncles.len() == max_uncles_num {
+        loop {
+            if candidate_uncles.is_empty() || uncles.len() == max_uncles_num {
                 break;
             }
-
-            let epoch_number = current_epoch_ext.number();
-
-            // uncle must be same difficulty epoch with candidate
+            let (hash, block) = candidate_uncles
+                .pop_front()
```

Comment:
It doesn't insert back the poped uncle block to `candidate_uncles`, this uncle block will be excluded in the next `get_block_template` function invoking.